File: yuke/.yuke/tools/web_search.lua
-- Web search tool: Brave Search first, Tavily fallback. Both keys are read
-- from ~/.yuke/.env (BRAVE_API_KEY, TAVILY_API_KEY), falling back to the
-- process environment. Ported from the pi web_search extension at
-- github.com/xyaman/pi-config/extensions/web_search.

local home = yuke.env.get("HOME")

-- Today's date (YYYY-MM-DD) for the tool description, so the model anchors
-- "recent" to the real current year instead of defaulting to older ones. The OS
-- stdlib is dropped from the yuke VM, so this shells out once at load time.
local function today()
	local ok, r = pcall(yuke.exec, "date +%Y-%m-%d")
	if ok and r and r.stdout then
		return (r.stdout:gsub("%s+$", ""))
	end
	return ""
end

-- Percent-encode a query-string value (RFC 3986 unreserved set kept literal).
local function urlencode(s)
	return (tostring(s):gsub("[^%w%-._~]", function(c)
		return string.format("%%%02X", string.byte(c))
	end))
end

-- Parse a simple .env file: skip comments/blank lines, split on first `=`,
-- strip surrounding quotes from the value.
local function parse_env(text)
	local env = {}
	for line in (text or ""):gmatch("[^\r\n]+") do
		local s = line:match("^%s*(.-)%s*$")
		if s ~= "" and s:sub(1, 1) ~= "#" then
			local eq = s:find("=", 1, true)
			if eq then
				local key = s:sub(1, eq - 1):match("^%s*(.-)%s*$")
				local value = s:sub(eq + 1):match("^%s*(.-)%s*$")
				if (value:sub(1, 1) == '"' and value:sub(-1) == '"')
					or (value:sub(1, 1) == "'" and value:sub(-1) == "'")
				then
					value = value:sub(2, -2)
				end
				env[key] = value
			end
		end
	end
	return env
end

-- Load Brave/Tavily keys from ~/.yuke/.env, then the process environment.
local function load_keys()
	local keys = {}
	local path = home and (home .. "/.yuke/.env") or nil
	if path and yuke.fs.exists(path) then
		local ok, text = pcall(yuke.fs.read, path)
		if ok and text then
			local env = parse_env(text)
			keys.brave = env.BRAVE_API_KEY
			keys.tavily = env.TAVILY_API_KEY
		end
	end
	keys.brave = keys.brave or yuke.env.get("BRAVE_API_KEY")
	keys.tavily = keys.tavily or yuke.env.get("TAVILY_API_KEY")
	return keys
end

-- Run a Brave Web Search. Returns nil on any failure so the caller can fall
-- back to Tavily; rethrows only on cancellation (surfaced by yuke.http).
local function search_brave(query, max, key)
	local url = "https://api.search.brave.com/api/v1/web/search?q="
		.. urlencode(query) .. "&count=" .. max
	local ok, res = pcall(yuke.http.get, url, {
		headers = { ["X-Subscription-Token"] = key, ["Accept"] = "application/json" },
		timeout_ms = 15000,
	})
	if not ok or res == nil or res.status ~= 200 then
		return nil
	end
	local ok2, data = pcall(yuke.json.decode, res.body)
	if not ok2 or type(data) ~= "table" then
		return nil
	end
	local results = {}
	for _, r in ipairs((data.web and data.web.results) or {}) do
		results[#results + 1] = { title = r.title, url = r.url, description = r.description }
	end
	return results
end

-- Run a Tavily search. Last resort: throws on any failure.
local function search_tavily(query, max, key)
	local res = yuke.http.post("https://api.tavily.com/search", {
		headers = { ["Content-Type"] = "application/json" },
		json = { api_key = key, query = query, max_results = max, search_depth = "basic" },
		timeout_ms = 15000,
	})
	if res == nil or res.status ~= 200 then
		local status = (res and res.status) or "?"
		local body = (res and res.body) or ""
		error(string.format("Tavily API error (%s): %s", status, body))
	end
	local data = yuke.json.decode(res.body)
	local results = {}
	for _, r in ipairs(data.results or {}) do
		results[#results + 1] = { title = r.title, url = r.url, description = r.content }
	end
	return results
end

-- Format results for the model: a header line per provider plus numbered entries.
local function format_results(query, results, provider, requested_max)
	local lines = {
		string.format('Query: "%s"', query),
		string.format("Provider: %s", provider),
		string.format("Results: %d (requested %d)", #results, requested_max),
		"",
	}
	if #results == 0 then
		lines[#lines + 1] = "No results found."
	else
		for i, r in ipairs(results) do
			lines[#lines + 1] = string.format("[%d] %s\nURL: %s\n%s", i, r.title, r.url, r.description)
			lines[#lines + 1] = ""
		end
	end
	return table.concat(lines, "\n")
end

yuke.tool({
	name = "web_search",
	description = "Search the web for information beyond the knowledge cutoff: recent events, library/API versions, or facts that may have changed. Returns results with title, URL, and a snippet. Account for Today's date in <env> — use the current year, not an older one, in queries for recent docs. `max_results` defaults to 5 (max 10).",
	params = { query = "string", max_results = "integer?" },
	handler = function(args)
		local query = args.query
		local max = args.max_results or 5
		max = math.max(1, math.min(10, max))
		local keys = load_keys()

		local results = nil
		local provider = "tavily"

		if keys.brave then
			results = search_brave(query, max, keys.brave)
			if results then
				provider = "brave"
			end
		end

		if not results then
			if not keys.tavily then
				error("Brave Search failed (or is not configured) and no TAVILY_API_KEY is set in ~/.yuke/.env. Both search providers are unavailable.")
			end
			results = search_tavily(query, max, keys.tavily)
			provider = "tavily"
		end

		return format_results(query, results, provider, max)
	end,
})

return true

File: yuke/.config/yuke/tools/web_search.js
import { exec } from "yuke:exec";

const BRAVE_URL = "https://api.search.brave.com/res/v1/web/search";
const TAVILY_URL = "https://api.tavily.com/search";
const REQUEST_TIMEOUT_MS = 20000;
const MAX_QUERY_CHARS = 400;
const MAX_TITLE_CHARS = 160;
const MAX_URL_CHARS = 2048;
const MAX_SNIPPET_CHARS = 360;
const MAX_OUTPUT_CHARS = 4096;

function shellQuote(value) {
  return "'" + String(value).replaceAll("'", "'\\''") + "'";
}

function keyPrelude(name) {
  return [
    "key=\"\"",
    'config_home=${XDG_CONFIG_HOME:-$HOME/.config}',
    'env_file="$config_home/yuke/.env"',
    'if [ -r "$env_file" ]; then',
    `  key=$(sed -n 's/^[[:space:]]*${name}[[:space:]]*=[[:space:]]*//p' "$env_file" | head -n 1 | sed -e 's/^\"//' -e 's/\"$//' -e "s/^'//" -e "s/'$//")`,
    "fi",
    `key=\${key:-\$${name}}`,
    '[ -n "$key" ] || exit 2',
  ].join("\n");
}

async function curl(command, signal) {
  const result = await exec(command, { timeoutMs: REQUEST_TIMEOUT_MS, signal });
  if (result.code !== 0 || result.signal !== null || result.timedOut || result.stdoutDropped !== 0) return null;
  return result.stdout;
}

function compactText(value, limit) {
  if (typeof value !== "string") return "";
  const text = value.slice(0, limit * 4).replace(/\s+/g, " ").trim();
  if (text.length <= limit) return text;

  let end = limit - 1;
  if (end > 0 && text.charCodeAt(end - 1) >= 0xd800 && text.charCodeAt(end - 1) <= 0xdbff) end -= 1;
  return text.slice(0, end) + "…";
}

function compactResults(items, snippetField) {
  const results = [];
  const urls = new Set();
  for (const item of items) {
    if (item === null || typeof item !== "object" || typeof item.url !== "string") continue;
    const url = item.url.trim();
    if (url.length === 0 || url.length > MAX_URL_CHARS || !/^https?:\/\/\S+$/i.test(url) || urls.has(url)) continue;

    urls.add(url);
    results.push({
      title: compactText(item.title, MAX_TITLE_CHARS) || url,
      description: compactText(item[snippetField], MAX_SNIPPET_CHARS),
      url,
    });
  }
  return results;
}

function parseResults(body, container, snippetField) {
  try {
    const data = JSON.parse(body);
    const source = container === null ? data : data[container];
    if (source === null || typeof source !== "object" || !Array.isArray(source.results)) return null;
    return compactResults(source.results, snippetField);
  } catch {
    return null;
  }
}

async function searchBrave(query, maxResults, signal) {
  const command = `${keyPrelude("BRAVE_API_KEY")}
curl --silent --show-error --fail --connect-timeout 5 --max-time 15 -G \\
  -H "X-Subscription-Token: $key" \\
  -H "Accept: application/json" \\
  --data-urlencode ${shellQuote(`q=${query}`)} \\
  --data ${shellQuote(`count=${maxResults}`)} \\
  ${shellQuote(BRAVE_URL)}`;
  const body = await curl(command, signal);
  return body === null ? null : parseResults(body, "web", "description");
}

async function searchTavily(query, maxResults, signal) {
  const queryJson = JSON.stringify(query);
  const command = `${keyPrelude("TAVILY_API_KEY")}
body=$(printf '{"query":%s,"max_results":%d,"search_depth":"basic","include_answer":false,"include_raw_content":false,"include_images":false}' ${shellQuote(queryJson)} ${maxResults})
curl --silent --show-error --fail --connect-timeout 5 --max-time 15 \\
  -H "Authorization: Bearer $key" \\
  -H "Content-Type: application/json" \\
  --data-raw "$body" \\
  ${shellQuote(TAVILY_URL)}`;
  const body = await curl(command, signal);
  return body === null ? null : parseResults(body, null, "content");
}

function formatResults(results) {
  if (results.length === 0) return "No sources found.";

  const lines = ["Sources:"];
  let length = lines[0].length;
  let shown = 0;
  for (const [index, item] of results.entries()) {
    const entry = `${index + 1}. ${item.title} — ${item.url}${item.description === "" ? "" : `\n   ${item.description}`}`;
    if (length + 1 + entry.length > MAX_OUTPUT_CHARS) break;
    lines.push(entry);
    length += 1 + entry.length;
    shown += 1;
  }
  if (shown < results.length) lines.push(`… ${results.length - shown} more source${results.length - shown === 1 ? "" : "s"} omitted.`);
  return lines.join("\n");
}

function parseArgs(args) {
  if (args === null || typeof args !== "object" || Array.isArray(args)) throw new Error("The arguments must be an object.");
  for (const name of Object.keys(args)) if (name !== "query" && name !== "max_results") throw new Error(`Unknown argument: ${name}`);
  if (typeof args.query !== "string") throw new Error("query must be a nonempty string.");
  const query = args.query.trim();
  if (query === "") throw new Error("query must be a nonempty string.");
  if (query.length > MAX_QUERY_CHARS) throw new Error(`query must contain at most ${MAX_QUERY_CHARS} characters.`);

  const maxResults = args.max_results === undefined ? 5 : args.max_results;
  if (typeof maxResults !== "number" || !Number.isSafeInteger(maxResults) || maxResults < 1 || maxResults > 10) {
    throw new Error("max_results must be an integer from 1 to 10.");
  }
  return { query, maxResults };
}

export const webSearchPlugin = {
  name: "web-search",
  apply(ctx) {
    ctx.tools.define({
      name: "web_search",
      description: "Search the web for recent events, library and API versions, and facts that may have changed. Results contain compact title, URL, and snippet sources. Use the current year for recent information.",
      parameters: {
        type: "object",
        properties: {
          query: { type: "string", description: "The web search query, up to 400 characters." },
          max_results: { type: "integer", minimum: 1, maximum: 10, description: "The number of sources to return. Defaults to 5." },
        },
        required: ["query"],
        additionalProperties: false,
      },
      execute: async (rawArgs, signal) => {
        const { query, maxResults } = parseArgs(rawArgs);
        const brave = await searchBrave(query, maxResults, signal);
        if (brave !== null) return formatResults(brave);

        const tavily = await searchTavily(query, maxResults, signal);
        if (tavily !== null) return formatResults(tavily);
        throw new Error("Both search providers failed. Set BRAVE_API_KEY or TAVILY_API_KEY in ~/.config/yuke/.env or the process environment.");
      },
    });
  },
};
